File: deno/day05.ts
import {DaySolution, parseData} from './_tools.ts';
import {parseLines} from './_helpers.ts';

type Line = number[];

class Data {
  public rules = new Map<number, Map<number, number>>();
  public lines = [] as Line[];

  addRule(s: string) {
    const [a, b] = s.split("|").map((i) => parseInt(i));

    if (!this.rules.has(a)) this.rules.set(a, new Map());
    this.rules.get(a)!.set(b, 1);

    if (!this.rules.has(b)) this.rules.set(b, new Map());
    this.rules.get(b)!.set(a, -1);
  }

  addLine(s: string) {
    this.lines.push(s.split(",").map((i) => parseInt(i)));
  }

  checkRule(a: number, b: number): number {
    return this.rules.get(a)?.get(b) ?? 0;
  }

  isSorted(line: Line): boolean {
    return line.every((v, i) => i === 0 || this.checkRule(line[i - 1], v) >= 0);
  }

  getLineRank(line: Line): number {
    return line[Math.floor(line.length / 2)];
  }
}

export class Day05 implements DaySolution<Data, number> {
  day = 5;

  parseData(dataLabel: "Example" | "Input"): Data {
    let readRules = true;
    return parseLines<string>(
      parseData(this.day, dataLabel),
      (line) => line,
      false,
    )
      .reduce((data, line) => {
        if (line === "") {
          readRules = false;
          return data;
        }
        if (readRules) {
          data.addRule(line);
        } else {
          data.addLine(line);
        }
        return data;
      }, new Data());
  }

  part1(input: Data): number {
    return input.lines
      .filter((line) => input.isSorted(line))
      .map((line) => input.getLineRank(line))
      .reduce((res, r) => res + r, 0);
  }

  part2(input: Data): number {
    return input.lines
      .filter((line) => !input.isSorted(line))
      .map(line => line.sort((a, b) => input.checkRule(a, b)))
      .map((line) => input.getLineRank(line))
      .reduce((res, r) => res + r, 0);
  }
}

File: rust/src/day05.rs
use crate::tools::read_raw_data;
use crate::DaySolution;
use std::cmp::Ordering;
use std::collections::HashMap;

struct SortRule(HashMap<(i32, i32), Ordering>);
impl SortRule {
    fn new() -> Self {
        SortRule(HashMap::<(i32, i32), Ordering>::new())
    }

    fn add_rule(&mut self, str: &str) {
        let v = str.split("|").map(|s| { s.parse::<i32>().unwrap() }).collect::<Vec<i32>>();
        self.0.insert((v[0], v[1]), Ordering::Less);
        self.0.insert((v[1], v[0]), Ordering::Greater);
    }

    fn get_rule(&self, a: i32, b: i32) -> Ordering {
        *self.0.get(&(a, b)).unwrap_or(&Ordering::Equal)
    }

    fn is_sorted(&self, list: &Vec<i32>) -> bool {
        list.windows(2).all(|w| { self.get_rule(w[0], w[1]) == Ordering::Less })
    }
}

pub struct Data(SortRule, Vec<Vec<i32>>);

pub struct Day05;

impl DaySolution<Data, i32> for Day05 {
    fn new() -> Self {
        Day05 {}
    }

    fn day(&self) -> u16 {
        5
    }

    fn parse_input(&mut self, example: bool) -> Data {
        let raw_data = read_raw_data(self.day(), example);
        let mut read_rules = true;
        let mut data = Data(SortRule::new(), Vec::new());
        raw_data.lines().for_each(|line| {
            if line.trim().is_empty() {
                read_rules = false;
                return;
            }

            if read_rules {
                data.0.add_rule(line);
            } else {
                data.1.push(line.split(",").map(|s| { s.parse::<i32>().unwrap() }).collect());
            }
        });
        data
    }

    fn part1(&mut self, input: &Data) -> i32 {
        let rules = &input.0;
        let lists = &input.1;
        lists.iter()
            .filter(|list| { rules.is_sorted(list) })
            .map(|list| { list[list.len() / 2] })
            .sum()
    }

    fn part2(&mut self, input: &Data) -> i32 {
        let rules = &input.0;
        let lists = &input.1;
        lists.iter()
            .filter(|list| { !rules.is_sorted(list) })
            .map(|list| {
                let mut sorted = list.clone();
                sorted.sort_by(|a, b| { rules.get_rule(*a, *b) });
                sorted[sorted.len() / 2]
            })
            .sum()
    }
}
